src/codegen/debug.rs

```diff
@@ -177,16 +169,33 @@ impl<'ink> DebugBuilderEnum<'ink> {
         root: Option<&Path>,
         optimization: OptimizationLevel,
         debug_level: DebugLevel,
+        output_format: FormatOption,
     ) -> Self {
-        let dwarf_version: BasicMetadataValueEnum<'ink> = context.i32_type().const_int(5, false).into();
         match debug_level {
             DebugLevel::None => DebugBuilderEnum::None,
-            DebugLevel::VariablesOnly | DebugLevel::Full => {
+            DebugLevel::VariablesOnly(version) | DebugLevel::Full(version) => {
+                let dwarf_version: BasicMetadataValueEnum<'ink> =
+                    context.i32_type().const_int(version as u64, false).into();
                 module.add_metadata_flag(
                     "Dwarf Version",
                     inkwell::module::FlagBehavior::Warning,
                     context.metadata_node(&[dwarf_version]),
                 );
+                // when emitting IR, `LLVMParseIRInContext` expects "Debug Info Version" metadata, with the specified version
+                // matching the LLVM version or otherwise it will emit a warning and strip DI from the IR.
+                // These metadata flags are not mutually exclusive.
+                if matches!(output_format, FormatOption::IR) {
```

Comment:
Why not do this for all cases?

Reply:
That's a good question. Since the warning only occured when emitting IR, I only added this when expecting IR output. I have checked the clang [sources ](https://github.com/llvm/llvm-project/blob/64a849a52e08827e889be22ed3ceafe62cd03793/clang/lib/CodeGen/CodeGenModule.cpp#L1003C1-L1008C61)and it looks like they add this metadata whenever any debug info is present in a module.